lib/services/notifications_cache_test.go

```diff
@@ -0,0 +1,457 @@
+	})
+	require.NoError(t, err)
+
+	notificationService, err := local.NewNotificationsService(bk, clock)
+	require.NoError(t, err)
+
+	svcs := notificationServices{
+		Events:        local.NewEventsService(bk),
+		Notifications: notificationService,
+	}
+
+	cache, err := services.NewUserNotificationCache(services.NotificationCacheConfig{
+		Events: svcs,
+		Getter: svcs,
+	})
+	require.NoError(t, err)
+
+	return svcs, cache
+}
+
+func newGlobalNotificationPack(t *testing.T) (notificationServices, *services.GlobalNotificationCache) {
+	ctx := context.Background()
+	clock := clockwork.NewFakeClock()
+	bk, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clock,
+	})
+	require.NoError(t, err)
+
+	notificationService, err := local.NewNotificationsService(bk, clock)
+	require.NoError(t, err)
+
+	svcs := notificationServices{
+		Events:        local.NewEventsService(bk),
+		Notifications: notificationService,
+	}
+
+	cache, err := services.NewGlobalNotificationCache(services.NotificationCacheConfig{
+		Events: svcs,
+		Getter: svcs,
+	})
+	require.NoError(t, err)
+
+	return svcs, cache
+}
+
+// TestUserNotificationsCache verifies the expected behaviors of the user-specific notifications cache.
+func TestUserNotificationsCache(t *testing.T) {
+	t.Parallel()
+
+	svcs, cache := newUserNotificationPack(t)
+	defer cache.Close()
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// To test that the streaming of notifications for a user is correct, we will have the mock list of user-specific notifications
+	// contain notifications for two users: alice and bob.
+	usernameA := "alice"
+	usernameB := "bob"
+
+	// Describe a set of mock user notifications to be created.
+	// The number in the description represents the order in which they are created, eg. alice-1 is the first (and thus oldest) notification to be created for alice.
+	testNotifications := []struct {
+		username    string
+		description string
+	}{
+		{
+			username:    usernameB,
+			description: "bob-1",
+		},
+		{
+			username:    usernameB,
+			description: "bob-2",
+		},
+		{
+			username:    usernameA,
+			description: "alice-1",
+		},
+		{
+			username:    usernameB,
+			description: "bob-3",
+		},
+		{
+			username:    usernameB,
+			description: "bob-4",
+		},
+		{
+			username:    usernameA,
+			description: "alice-2",
+		},
+		{
+			username:    usernameA,
+			description: "alice-3",
+		},
+		{
+			username:    usernameA,
+			description: "alice-4",
+		},
+		{
+			username:    usernameB,
+			description: "bob-5",
+		},
+		{
+			username:    usernameA,
+			description: "alice-5",
+		},
+	}
+
+	// We want to keep the notification ID's of the created notifications for later to test deleting notifications.
+	notificationIdsA := []string{}
+	notificationIdsB := []string{}
+
+	// Create the user-specific notifications, with a half second delay between each one.
+	for _, n := range testNotifications {
+		time.Sleep(500 * time.Millisecond)
```

Comment:
I ran into some flakiness issues with other notifications tests because when the test notifications were created in quick succession during code execution, the timestamps in their UUID (which we use for sorting by date) were sometimes the same and so the lexicographical sorting would end up being based on the randomly generated portion of the UUID. 500 milliseconds is excessive though, I'll lower this delay